app/controllers/PromoLandingPage.scala

```diff
@@ -48,15 +49,20 @@ object PromoLandingPage extends Controller {
     implicit val promoCode = PromoCode(promoCodeStr)
     (for {
       promotion <- tpBackend.promoService.findPromotion(promoCode)
-      html <- getNewspaperLandingPage(promotion) orElse getDigitalPackLandingPage(promotion) orElse getWeeklyLandingPage(promotion)
+      html <- getLandingPage(promotion)
     } yield Ok(html)).getOrElse(Redirect("/" ? (internalCampaignCode -> s"FROM_P_${promoCode.get}")))
   }
+  def getLandingPage(promotion: AnyPromotion)(implicit promoCode: PromoCode): Option[Html] = {
+    getNewspaperLandingPage(promotion) orElse getDigitalPackLandingPage(promotion) orElse getWeeklyLandingPage(promotion)
+  }
 
   def preview() = GoogleAuthenticatedStaffAction { implicit request =>
+    val previewReturnHeaders = HandleXFrameOptionsOverrideHeader.HEADER_KEY -> s"ALLOW-FROM ${Config.previewXFrameOptionsOverride}"
     Form(Forms.single("promoJson" -> Forms.text)).bindFromRequest().fold(_ => NotFound, { jsString =>
-      Json.fromJson[AnyPromotion](Json.parse(jsString)).asOpt.flatMap(_.asDigitalPack)
-        .map(p => views.html.promotion.digitalpackLandingPage(edition, catalog, p.codes.headOption.getOrElse(PromoCode("")), p, PegdownMarkdownRenderer))
-        .fold[Result](NotFound)(p => Ok(p).withHeaders(HandleXFrameOptionsOverrideHeader.HEADER_KEY -> s"ALLOW-FROM ${Config.previewXFrameOptionsOverride}"))
+      Json.fromJson[AnyPromotion](Json.parse(jsString)).asOpt.flatMap { promotion =>
+        implicit val promoCode: PromoCode = promotion.codes.headOption.getOrElse(PromoCode(""))
+        getLandingPage(promotion)
```

Comment:
I think something like:

promotion.codes.headOption.flatMap { implicit x => getLandingPage(promotion) } might be better rather than creating an empty PromoCode. Plus then as all the flatMaps are nested, you could turn the whole thing into a for comprehension.